File: 3748.count-stable-subarrays.py
#
# @lc app=leetcode id=3748 lang=python3
#
# [3748] Count Stable Subarrays
#

# @lc code=start
from typing import List


class Solution:
    def countStableSubarrays(self, nums: List[int], queries: List[List[int]]) -> List[int]:
        n = len(nums)
        if n == 0:
            return []
        left = [0] * n
        for i in range(1, n):
            if nums[i - 1] <= nums[i]:
                left[i] = left[i - 1]
            else:
                left[i] = i
        prefix = [0] * (n + 1)
        for i in range(n):
            prefix[i + 1] = prefix[i] + left[i]
        
        def range_sum(l: int, r: int) -> int:
            def s(x: int) -> int:
                return x * (x + 1) // 2
            return s(r) - s(l - 1)
        
        ans = []
        for l, r in queries:
            m = r - l + 1
            triangle = m * (m + 1) // 2
            # Binary search for j0: smallest j >= l with left[j] > l
            lo = l
            hi = r + 1
            while lo < hi:
                mid = (lo + hi) // 2
                if left[mid] > l:
                    hi = mid
                else:
                    lo = mid + 1
            j0 = lo
            if j0 > r:
                res = triangle
            else:
                num_bad = r - j0 + 1
                sum_left_bad = prefix[r + 1] - prefix[j0]
                sum_max = l * (m - num_bad) + sum_left_bad
                sum_jp1 = range_sum(l, r) + m
                res = sum_jp1 - sum_max
            ans.append(res)
        return ans

# @lc code=end

File: 3748.count-stable-subarrays.cpp
#
# @lc app=leetcode id=3748 lang=cpp
#
# [3748] Count Stable Subarrays
#

# @lc code=start
class Solution {
public:
    vector<long long> countStableSubarrays(vector<int>& nums, vector<vector<int>>& queries) {
        int n = nums.size();
        
        // Assign segment id to each position
        // A new segment starts when nums[i] < nums[i-1] (non-decreasing segments)
        vector<int> seg_id(n);
        seg_id[0] = 0;
        for (int i = 1; i < n; i++) {
            seg_id[i] = seg_id[i-1] + (nums[i] < nums[i-1] ? 1 : 0);
        }
        
        int m = seg_id[n-1] + 1;  // Number of segments
        
        // Find segment boundaries
        vector<int> seg_start(m), seg_end(m);
        for (int i = 0; i < n; i++) {
            int s = seg_id[i];
            if (i == 0 || seg_id[i] != seg_id[i-1]) {
                seg_start[s] = i;
            }
            seg_end[s] = i;
        }
        
        // Compute prefix contributions
        // For a segment of length len, number of subarrays = len*(len+1)/2
        vector<long long> prefix_contrib(m + 1, 0);
        for (int j = 0; j < m; j++) {
            long long len = seg_end[j] - seg_start[j] + 1;
            prefix_contrib[j + 1] = prefix_contrib[j] + len * (len + 1) / 2;
        }
        
        // Process queries
        vector<long long> ans;
        ans.reserve(queries.size());
        for (const auto& q : queries) {
            int l = q[0], r = q[1];
            int a = seg_id[l], b = seg_id[r];
            
            if (a == b) {
                // l and r are in the same segment
                long long len = r - l + 1;
                ans.push_back(len * (len + 1) / 2);
            } else {
                // Different segments
                // Partial contribution from segment a (from l to seg_end[a])
                long long len_a = seg_end[a] - l + 1;
                // Partial contribution from segment b (from seg_start[b] to r)
                long long len_b = r - seg_start[b] + 1;
                // Full contributions from segments a+1 to b-1
                long long result = len_a * (len_a + 1) / 2;
                result += prefix_contrib[b] - prefix_contrib[a + 1];
                result += len_b * (len_b + 1) / 2;
                ans.push_back(result);
            }
        }
        
        return ans;
    }
};
# @lc code=end